File: src/main/java/net/lazars/garden/service/SunriseSunsetService.java
package net.lazars.garden.service;

import jakarta.annotation.PostConstruct;
import java.io.IOException;
import java.time.Instant;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import net.lazars.garden.client.SunriseSunsetClient;
import net.lazars.garden.model.SunriseSunsetResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import retrofit2.Response;

@Slf4j
@Service
@RequiredArgsConstructor
public class SunriseSunsetService {

  private final SunriseSunsetClient sunriseSunsetClient;

  @Value("${garden.location.latitude}")
  private String latitude;

  @Value("${garden.location.longitude}")
  private String longitude;

  @Getter @Setter private Instant sunsetData;

  @PostConstruct
  void init() {
    log.info("Initializing sunrise/sunset service");
    loadSunsetData("today");
    if (Instant.now().isAfter(sunsetData)) {
      loadUpcomingSunsetData();
    }
  }

  public Instant loadUpcomingSunsetData() {
    loadSunsetData("tomorrow");
    return sunsetData;
  }

  private void loadSunsetData(final String date) {
    final Response<SunriseSunsetResponse> response;
    try {
      response = sunriseSunsetClient.getSunsetData(latitude, longitude, date, 0).execute();
    } catch (IOException e) {
      log.error("Error loading sunrise/sunset data", e);
      return;
    }

    if (!response.isSuccessful()) {
      log.error("Error loading sunrise/sunset data: {}", response);
      return;
    }

    final SunriseSunsetResponse sunriseSunsetResponse = response.body();
    if (sunriseSunsetResponse != null) {
      final String sunset = sunriseSunsetResponse.getResults().getNauticalTwilightEnd();
      sunsetData = Instant.parse(sunset);
      log.info("Loaded sunset data: {}", sunsetData);
    } else {
      log.error("Error loading sunrise/sunset data: {}", response);
    }
  }
}

File: src/main/kotlin/net/lazars/garden/service/SunriseSunsetService.kt
package net.lazars.garden.service

import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.http.*
import kotlinx.coroutines.runBlocking
import net.lazars.garden.getHttpClient
import net.lazars.garden.model.SunriseSunsetResponse
import org.slf4j.LoggerFactory
import java.time.Instant

class SunriseSunsetService(private val latitude: String, private val longitude: String) {

    private val log = LoggerFactory.getLogger(javaClass)
    lateinit var sunsetData: Instant private set

    init {
        loadSunsetData("today")
        if (Instant.now().isAfter(sunsetData)) loadSunsetData("tomorrow")
    }

    fun loadUpcomingSunsetData(): Instant {
        loadSunsetData("tomorrow")
        return sunsetData
    }

    private fun loadSunsetData(date: String) {
        getHttpClient().use { client ->
            runBlocking {
                val response = client.get("https://api.sunrise-sunset.org/") {
                    url {
                        appendPathSegments("json")
                        parameters.append("lat", latitude)
                        parameters.append("lng", longitude)
                        parameters.append("date", date)
                        parameters.append("formatted", "0")
                        contentType(ContentType.Application.Json)
                    }
                }

                if (response.status == HttpStatusCode.OK) {
                    val responseData: SunriseSunsetResponse = response.body()
                    sunsetData = Instant.parse(responseData.results.twilightEnd)
                    log.info("Loaded sunset data: $sunsetData")
                }
            }
        }
    }
}
